### ISSUE TITLE
`exec=off` has no effect with `mountpoint=legacy` (on Linux)

### ISSUE THREAD

**Issue by pacien** [openzfs/zfs#12852]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| NixOS
Distribution Version	| 21.11
Kernel Version	| Linux 5.10.81
Architecture	| x86_amd64
OpenZFS Version	| zfs-2.1.1-1, zfs-kmod-2.1.1-1

### Describe the problem you're observing

`exec=off` has no effect with `mountpoint=legacy`.

In the presence of the former option set to `off`, the dataset should be
mounted with the `noexec` mount option. This is not the case.

### Describe how to reproduce the problem

```sh
zfs create zpool/reference -o exec=off -o mountpoint=/mnt/reference
mount
# zpool/reference on /mnt/reference type zfs (rw,noexec,noatime,xattr,noacl)
#                                                ^ present

zfs create zpool/execoff_legacy -o exec=off -o mountpoint=legacy
mkdir /mnt/execoff_legacy
mount -t zfs zpool/execoff_legacy /mnt/execoff_legacy
mount
# zpool/execoff_legacy on /mnt/execoff_legacy type zfs (rw,relatime,xattr,noacl)
#                                                       ^ noexec absent
```

### Include any warning/errors/backtraces from the system logs

_None._

**Comment by pacien** [comment]
I tried to reproduce the issue with different environments.
Here are the results:

OS | ZFS version | Test result
 --- | --- | ---
NixOS 21.11	(kernel 5.10.81, util-linux mount) | 2.1.1-1 | Issue reproduced
NixOS 21.11	(kernel 5.10.81, busybox mount) | 2.1.1-1 | Issue reproduced
Debian 11.1 (kernel 5.10.0-9, util-linux mount) | 2.0.3-9 | Issue reproduced
Debian 11.1 (kernel 5.10.0-9, busybox mount) | 2.0.3-9 | Issue reproduced
FreeBSD 13.0-RELEASE-p5 | 2.0.0-FreeBSD_gf11b09dec | Issue __not__ present

It seems that this issue only affects Linux and probably not BSDs.

The issue is also happening with both the util-linux and busybox
implementations of mount.

**Comment by h1z1** [comment]
Don't see how that is a defect other then fbsd doing it _wrong_.  Legacy is supposed to leave flags upto the os/fstab.
As per mangpage:


> If a file system's mount point is set to  legacy, ZFS makes no attempt
> to manage the file system, and the administrator is responsible for 
> mounting and unmounting the file system.
```

**Comment by pacien** [comment]
[zfsprops(7/8)] explicitly states:

```
   Temporary Mount Point Properties
     When a file system is mounted, either through mount(8) for legacy
     mounts or the zfs mount command for normal file systems, its mount op‐
     tions are set according to its properties.  The correlation between
     properties and mount options is as follows:
           atime         atime/noatime
           canmount      auto/noauto
           devices       dev/nodev
           exec          exec/noexec
           readonly      ro/rw
           relatime      relatime/norelatime
           setuid        suid/nosuid
           xattr         xattr/noxattr
           nbmand        mand/nomand
           context=      context=
           fscontext=    fscontext=
           defcontext=   defcontext=
           rootcontext=  rootcontext=

     In addition, these options can be set on a per-mount basis using the
     -o option, without affecting the property that is stored on disk.  The
     values specified on the command line override the values stored in the
     dataset.  The nosuid option is an alias for nodevices,nosetuid.  These
     properties are reported as "temporary" by the zfs get command.  If the
     properties are changed while the dataset is mounted, the new setting
     overrides any temporary settings.
```

[zfsprops(7/8)]: https://openzfs.github.io/openzfs-docs/man/7/zfsprops.7.html?highlight=legacy%20noexec#Temporary_Mount_Point_Properties

So this does look like a defect affecting Linux.

**Comment by h1z1** [comment]
Pedantic perhaps but I'd take that as different things..  The text is certainly confusing..  ZFS has no idea where to mount the filesystem given the property is overridden (mountpoint=legacy).  Thus to me at least, it means those options are passed from mount (userspace) TO zfs (kernel), not zfs to mount.

```
# zfs create tank/foo
# mkdir /tmp/foo
# zfs set mountpoint=/tmp/foo tank/foo
# grep /tmp/foo /proc/mounts              
tank/foo /tmp/foo zfs rw,noatime,xattr,noacl 0 0
# umount /tmp/foo
# grep /tmp/foo /proc/mounts              
# zfs set exec=off tank/foo
# grep /tmp/foo /proc/mounts              
# zfs mount /tmp/foo
cannot open '/tmp/foo': leading slash in name
# zfs mount tank/foo
# grep /tmp/foo /proc/mounts              
tank/foo /tmp/foo zfs rw,noexec,noatime,xattr,noacl 0 0
# umount /tmp/foo
# zfs set mountpoint=legacy tank/foo
# zfs mount tank/foo
cannot mount 'tank/foo': legacy mountpoint
use mount(8) to mount this filesystem
# mount -t zfs -v  tank/foo /tmp/foo
mount.zfs:
  dataset:    "tank/foo"
  mountpoint: "/tmp/foo"
  mountflags: 0x0
  zfsflags:   0x0
  mountopts:  "rw"
  mtabopts:   "rw"
#
# umount /tmp/foo
# zfs set readonly=on tank/foo
# mount -t zfs -v  tank/foo /tmp/foo
mount.zfs:
  dataset:    "tank/foo"
  mountpoint: "/tmp/foo"
  mountflags: 0x0
  zfsflags:   0x0
  mountopts:  "rw"
  mtabopts:   "rw"
# umount /tmp/foo
# mount -t zfs -v -o ro,noexec tank/foo /tmp/foo
mount.zfs:
  dataset:    "tank/foo"
  mountpoint: "/tmp/foo"
  mountflags: 0x9
  zfsflags:   0x0
  mountopts:  "ro,noexec"
  mtabopts:   "ro,noexec"

```

tl;dr it's doing what I'd expect.  The documentation is incorrect or at the very least inconsistent.

**Comment by Dragon-debian** [comment]
I get the same results on debian 11 ( bullseye), both with `busybox` mount and the mount command provided by `mount` package.
setting `xattr` and `acltype` properties, affects the dataset even when the `mountpoint` property is set to `legacy`
`setuid`  have `devices` no affect on datasets with `mountpoint=legacy`, just like `exec`

**Comment by pacien** [comment]
Still an issue.

**Comment by amotin** [comment]
This looks similar to what was addressed by #16646.  If you use third-party `mount` binary, it won't set the mount properties as done by `mount.zfs`, called by "normal" `mount`.  I can not propose anything better than using proper `mount` tool, or directly calling `mount.zfs`.
